Improve smart contract security best practices (#722)
dfinity/docs: modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc

= Rust Canister Development Security Best Practices

> == Smart Contracts under Canister Control

Author: Sure, adapted to "Smart Contract Canister Control"

==== Recommendation

> * Consider passing canister control to a decentralized governance system such as the Internet Computer's Service Nervous System (SNS), so that changes to the canister must be proposed and voted upon. If the SNS is used, use an NNS-blessed version running on the SNS subnet, meaning that SNS is under NNS control. The SNS will be available soon.

Author: Do we have any public resources (doc, blogs, ...) regarding the SNS that we could link here?

Reviewer: how about just pointing to the roadmap? This would give extra info on what we mean by "will be available soon" too and provide a minimal description of the SNSs

> * Consider passing canister control to a decentralized governance system such as the Internet Computer's Service Nervous System (SNS), so that changes to the canister must be proposed and voted upon. If the SNS is used, use an NNS-blessed version running on the SNS subnet, meaning that SNS is under NNS control. The SNS will be available soon.

Reviewer 2: I second Lara's changes.

> * Another option would be to create an immutable canister smart contract by removing the canister controller completely. However, note that this implies that the canister cannot be upgraded, which may have severe implications in case e.g. a bug were found. The option to use a decentralized governance system and thus being able to upgrade smart contracts is a big advantage of the Internet Computer ecosystem.

Reviewer: nit: now this part re: immutable cansiters is not fully in line with the title "Smart Contracts under Decentralized Governance Control".
I guess the more generic title would be "Canister Control" or so.

Author: Agreed. I adapted the title to "Smart Contracts under Canister Control"

> * Another option would be to create an immutable canister smart contract by removing the canister controller completely. However, note that this implies that the canister cannot be upgraded, which may have severe implications in case e.g. a bug were found. The option to use a decentralized governance system and thus being able to upgrade smart contracts is a big advantage of the Internet Computer ecosystem.

Reviewer: nit: " advantage of the Internet Computer ecosystem" compared to what?
(I assume here we mean "compared to other blockchains", but maybe we don't want to be that straightforward, so leaving it might be the better option?)

Author: I like "compared to other blockchains" and added this.

> * It is also possible to implement a DAO (link:https://en.wikipedia.org/wiki/Decentralized_autonomous_organization[Decentralized Autonomous Organization]) on the IC from scratch. If you decide to do this (e.g. along the lines of the link:https://smartcontracts.org/docs/samples/dao.html[basic DAO example]), be aware that this is security critical and must be security reviewed carefully.

Reviewer: should we add here that if this option is chosen, then one important aspect for users to check /verify is again the control over these DAOs (i.e., the DAO should control itself)

Author: Agreed, added "Furthermore, users will need to verify that the DAO is controlled by itself."

> * Note that, contrary to some other blockchains, smart contracts (also immutable ones) need cycles to run, and they can receive cycles.

Reviewer 2: Immutable canisters are related because the presumed goal of the developer is that _there is no centralized control_.

> * Note that, contrary to some other blockchains, smart contracts (also immutable ones) need cycles to run, and they can receive cycles.

Reviewer: right I see that the subject is related, but this bullet point seems to raise a point that people could be suspicious for _for immutable canisters_. If so, I think it should be a subpoint of the immutable canisters mentioned rather than another point.

Or stated differently: it is unclear to me why this information is in a bullet list of things that seems to list options for having non-central controllers for a canister (it does not add another option for this). And since I am confused about that I suspect that others might too. Therefore, I wonder if the reason can be made explicit or if it makes sense to move this bullet point to somewhere else

> * Note that, contrary to some other blockchains, smart contracts (also immutable ones) need cycles to run, and they can receive cycles.

Author: Thanks. Good point, this does not fit the structure, as the other points list options for creating smart contracts. This is a purely informational (and not really essential) point that relates to the immutable contracts. I added it as a sub point there, does that work for you? If you have a better idea on where to place this, please let me know.

=== Verify the ownership of smart contracts you depend on

==== Security Concern

> If a canister depends on a canister smart contract (i.e. makes inter-canister calls to it), it is essential that the canister smart contract is owned by a decentralized governance system. Otherwise, i.e. if it has a controller, they could modify the smart contract whenever they like, e.g. to steal assets held by the canister.

Reviewer: One more detail: an SNS can also change the canister whenever they like, but it is more controlled and maybe more visible.. 
Maybe instead of "whenever they like" we could rather say "in silence"/"without other noticing" or so?

> If a canister depends on a canister smart contract (i.e. makes inter-canister calls to it), it is essential that the canister smart contract is owned by a decentralized governance system. Otherwise, i.e. if it has a controller, they could modify the smart contract whenever they like, e.g. to steal assets held by the canister.

Author: Thanks, I made the suggested change. 

Also switched the wording to "without others noticing".

==== Recommendation

> If you interact with a canister that must be an smart contract, make sure it is controlled by the NNS, service nervous system (SNS) or a decentralized governance system, and review under what conditions and by whom the smart contract can be changed.

Reviewer 2: Perhaps better to say "a canister that you require to be decentralized" instead of "canister that must be a smart contract".
```suggestion
If you interact with a canister that you require to be decentralized, make sure it is controlled by the NNS, service nervous system (SNS) or a decentralized governance system, and review under what conditions and by whom the smart contract can be changed.
```

> If you interact with a canister that must be an smart contract, make sure it is controlled by the NNS, service nervous system (SNS) or a decentralized governance system, and review under what conditions and by whom the smart contract can be changed.

Author: Agreed and adapted accordingly.

> If you interact with a canister that you require to be decentralized, make sure it is controlled by the NNS, service nervous system (SNS) or a decentralized governance system, and review under what conditions and by whom the smart contract can be changed.

Reviewer: ```suggestion
If you interact with a canister that you require to be decentralized, make sure it is controlled by the NNS, a service nervous system (SNS) or a decentralized governance system, and review under what conditions and by whom the smart contract can be changed.
```

Author: Done
